pylint/checkers/variables.py

```diff
@@ -1528,10 +1532,9 @@ def _check_consumer(
                         )
                         and node.name in node.root().locals
                     ):
-                        self.add_message(
-                            "undefined-variable", args=node.name, node=node
-                        )
-                        return (VariableVisitConsumerAction.CONSUME, found_nodes)
+                        if defined_by_stmt:
+                            current_consumer.mark_as_consumed(node.name, [node])
+                        return (VariableVisitConsumerAction.CONTINUE, None)
```

Comment:
I'm probably missing something here, but why can't we return `Consume` with `[node]` instead of `found_nodes`?

Reply:
CONTINUE is the only action that will allow further messages to be emitted in the same scope (the original issue being fixed), but when leaving the scope, if it's a `defined_by_stmt` line we want to make sure the name is consumed and does not emit `undefined-variable`.

Otherwise, we would need a fourth VariableVisitConsumerAction action along the lines of CONSUME_AND_CONTINUE.

Reply:
Hm, what do you thinks makes most sense? `CONSUME_AND_CONTINUE` sounds the most future-proof to me? I feel like having a `mark_as_consumed` somewhere in this method makes the consumption harder to comprehend again.

Reply:
An alternative might be to just reduce the actions to CONTINUE and RETURN and then permit the second element of the tuple to have elements to mark as consumed no matter the first value?

Reply:
Hmm that's not bad actually. Performance wise it would be similar, as `if action == CONSUME` or `if consume_nodes` should be quite similar.
Perhaps that should be done in another PR though?

_Look at me thinking this PR showed how good and future-proof our previous code was 😅_

Reply:
Follow up PR sounds right.

Just making sure I follow: did you want to avoid adding a fourth consumer action for now?

Reply:
Yes, if we know we're going to reduce down to two actions in the foreseeable future we might as well keep it like this.